### PR TITLE
Shellcheck: test scripts

### PR DESCRIPTION
**Pull request by SkamDart**
# Motivation
This PR intends to add shellcheck fixes to a subset of the functional tests

# Context

Partial resolve of https://github.com/NixOS/nix/issues/10795.
pre-commit hook is kind of a pain to work with so this was also validated with the following:
```bash
 git diff origin/master --name-only | xargs shellcheck
```

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: tests/functional/binary-cache.sh [hunk 1]
```diff
@@ -25,49 +25,49 @@ for path in "${paths[@]}"; do
...
+nix store copy-log --from file://"$cacheDir" "$(nix-store -qd "$outPath")"^'*'
+nix log "$outPath" | grep FOO
 
 basicDownloadTests() {
     # No uploading tests bcause upload with force HTTP doesn't work.
 
     # By default, a binary cache doesn't support "nix-env -qas", but does
     # support installation.
-    clearStore
+   clearStore
```
Comment (Ericson2314) [inline tests/functional/binary-cache.sh:44]: indentation issue?

Comment (SkamDart) [reply tests/functional/binary-cache.sh]: Fortunately, this doesn't affect the test itself and shellcheck is still happy.

```bash
bash-5.2$ ./mk/run-test.sh tests/functional/binary-cache.sh
ran test tests/functional/binary-cache.sh... [PASS]
bash-5.2$ shellcheck tests/functional/binary-cache.sh
bash-5.2$ echo $?
0
```

I went ahead and rebased, fixed, and force pushed.

**Comment by Ericson2314** [comment]
@SuperSandro2000 We need a... super reviewer for this. Would you like to be the one?

File: tests/functional/flakes/absolute-attr-paths.sh [hunk 2]
```diff
@@ -14,6 +14,6 @@ cat > $flake1Dir/flake.nix <<EOF
 }
 EOF
 
-[ "$(nix eval --impure --json $flake1Dir#.x)" -eq 1 ]
-[ "$(nix eval --impure --json $flake1Dir#x)" -eq 2 ]
-[ "$(nix eval --impure --json $flake1Dir#.packages.$system.x)" -eq 2 ]
+[ "$(nix eval --impure --json "$flake1Dir"#.x)" -eq 1 ]
+[ "$(nix eval --impure --json "$flake1Dir"#x)" -eq 2 ]
+[ "$(nix eval --impure --json "$flake1Dir"#.packages."$system".x)" -eq 2 ]
```
Comment (Ericson2314) [inline tests/functional/flakes/absolute-attr-paths.sh:19]: how did that every work with the `#`!

Comment (Ericson2314) [reply tests/functional/flakes/absolute-attr-paths.sh:19]: (oh, @roberth says because no whitespace around the `#`.)

Comment (Ericson2314) [reply tests/functional/flakes/absolute-attr-paths.sh:19]: Speaking of which, I would love to combine things like this into one big `"...."`, but I am fine saving that for a separate PR if it's harder to automate (thank `shellcheck --fix` or whatever may have assisted with this).

File: tests/functional/binary-cache.sh [hunk 3]
```diff
@@ -14,9 +14,9 @@ clearStore
 clearCache
 outPath=$(nix-build dependencies.nix --no-out-link)
 
-nix copy --to file://$cacheDir $outPath
+nix copy --to file://"$cacheDir" "$outPath"
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:17]: I would include ``file://`` in the quote, too but I think that makes not difference.

File: tests/functional/binary-cache.sh [hunk 4]
```diff
@@ -46,28 +46,28 @@ basicDownloadTests() {
 
     nix-env --substituters "file://$cacheDir" -f dependencies.nix -qas \* | grep -- "---"
 
-    nix-store --substituters "file://$cacheDir" --no-require-sigs -r $outPath
+    nix-store --substituters "file://$cacheDir" --no-require-sigs -r "$outPath"
 
-    [ -x $outPath/program ]
+    [ -x "$outPath"/program ]
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:51]: same here, I usually quote the entire string unless there is a wildcard asterix.

I will now stop commenting this for all other occurrences.

File: tests/functional/binary-cache.sh [hunk 5]
```diff
@@ -107,131 +107,135 @@ fi
...
 
 # Test whether fallback works if a NAR is corrupted. This does require --fallback.
 clearStore
 
-mv $cacheDir/nar $cacheDir/nar2
-mkdir $cacheDir/nar
-for i in $(cd $cacheDir/nar2 && echo *); do touch $cacheDir/nar/$i; done
+mv "$cacheDir"/nar "$cacheDir"/nar2
+mkdir "$cacheDir"/nar
+for i in $(cd "$cacheDir"/nar2 && echo *); do touch "$cacheDir"/nar/"$i"; done
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:122]: I think shellcheck does not detect this, but we should probably use find here, too

File: tests/functional/binary-cache.sh [hunk 6]
```diff
@@ -107,131 +107,135 @@ fi
...
+(! nix store cat --store file://"$cacheDir" "$outPath"/foobar)
 
 
 # Test NAR listing generation.
 clearCache
 
+
+# shellcheck disable=SC2016
+#
+# we're ignoring shellcheck because we really don't want shell expansion here
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:238]:
```suggestion
# preserve quotes variables in the single quoted string
# shellcheck disable=SC2016
```
I assume?

File: tests/functional/binary-cache.sh [hunk 7]
```diff
@@ -240,16 +244,20 @@ outPath=$(nix-build --no-out-link -E '
   }
 ')
 
-nix copy --to file://$cacheDir?write-nar-listing=1 $outPath
+nix copy --to file://"$cacheDir"?write-nar-listing=1 "$outPath"
 
+# shellcheck disable=2046
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:249]:
```suggestion
# shellcheck disable=SC2046
```

where do we have word splitting here? That's just subshells and redirect 🤔

File: tests/functional/binary-cache.sh [hunk 8]
```diff
@@ -240,16 +244,20 @@ outPath=$(nix-build --no-out-link -E '
...
+    <(jq -S < "$cacheDir"/$(basename "$outPath" | cut -c1-32).ls) \
     <(echo '{"version":1,"root":{"type":"directory","entries":{"bar":{"type":"regular","size":4,"narOffset":232},"link":{"type":"symlink","target":"xyzzy"}}}}' | jq -S)
 
 
 # Test debug info index generation.
 clearCache
 
+# shellcheck disable=SC2016
+#
+# we're ignoring shellcheck because we really don't want shell expansion here
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:260]:
```suggestion
# preserve quotes variables in the single quoted string
# shellcheck disable=SC2016
```
and so fort for the next ones

File: tests/functional/binary-cache.sh [hunk 9]
```diff
@@ -258,14 +266,17 @@ outPath=$(nix-build --no-out-link -E '
   }
 ')
 
-nix copy --to "file://$cacheDir?index-debug-info=1&compression=none" $outPath
+nix copy --to "file://$cacheDir?index-debug-info=1&compression=none" "$outPath"
 
 diff -u \
-    <(cat $cacheDir/debuginfo/02623eda209c26a59b1a8638ff7752f6b945c26b.debug | jq -S) \
+    <(jq -S < "$cacheDir"/debuginfo/02623eda209c26a59b1a8638ff7752f6b945c26b.debug) \
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:272]:
```suggestion
    <(jq -S "$cacheDir"/debuginfo/02623eda209c26a59b1a8638ff7752f6b945c26b.debug) \
```
Please test but jq should be able to read the file directly

File: tests/functional/binary-cache.sh [hunk 10]
```diff
@@ -275,22 +286,22 @@ expr='
...
-nix copy --to "file://$cacheDir" $outPath
+nix copy --to "file://$cacheDir" "$outPath"
 
 hashpart() {
   basename "$1" | cut -c1-32
 }
 
 # break the closure of out by removing doc
-rm $cacheDir/$(hashpart $docPath).narinfo
+rm "$cacheDir"/"$(hashpart "$docPath")".narinfo
```
Comment (SuperSandro2000) [inline tests/functional/binary-cache.sh:302]:
```suggestion
rm "$cacheDir/$(hashpart "$docPath")".narinfo
```
we don't need to unquote the /

File: tests/functional/brotli.sh [hunk 11]
```diff
@@ -9,15 +9,15 @@ cacheURI="file://$cacheDir?compression=br"
...
 clearCacheCache
 
-nix copy --from $cacheURI $outPath --no-check-sigs
+nix copy --from "$cacheURI" "$outPath" --no-check-sigs
 
-HASH2=$(nix hash path $outPath)
+HASH2=$(nix hash path "$outPath")
 
-[[ $HASH = $HASH2 ]]
+[[ $HASH = "$HASH2" ]]
```
Comment (SuperSandro2000) [inline tests/functional/brotli.sh:23]:
```suggestion
[[ $HASH == "$HASH2" ]]
```

File: tests/functional/ca/build-cache.sh [hunk 12]
```diff
@@ -26,7 +26,9 @@ copyAttr () {
     # Note: to copy CA derivations, we need to copy the realisations, which
     # currently requires naming the installables, not just the derivation output
     # path.
-    nix copy --to file://$cacheDir "${args[@]}"
+
+    # shellcheck disable=SC2154
```
Comment (SuperSandro2000) [inline tests/functional/ca/build-cache.sh:30]: I am a bit surprised by this because args is defined above

File: tests/functional/case-hack.sh [hunk 13]
```diff
@@ -4,18 +4,25 @@ source common.sh
 
 clearStore
 
-rm -rf $TEST_ROOT/case
+rm -rf "$TEST_ROOT"/case
 
 opts="--option use-case-hack true"
```
Comment (SuperSandro2000) [inline tests/functional/case-hack.sh:9]:
```suggestion
opts=("--option" "use-case-hack" "true")
```

File: tests/functional/case-hack.sh [hunk 14]
```diff
@@ -4,18 +4,25 @@ source common.sh
...
+
+# shellcheck disable=2086
+# putting quotes around "$opts" causes the test to fail with
+# 'unknown flag' on nix-store command because it is quoted
+nix-store $opts  --restore "$TEST_ROOT/case" < case.nar
+# shellcheck disable=2086
+nix-store $opts --dump "$TEST_ROOT/case" > "$TEST_ROOT/case.nar"
+cmp case.nar "$TEST_ROOT/case.nar"
+# shellcheck disable=2086
 [ "$(nix-hash $opts --type sha256 $TEST_ROOT/case)" = "$(nix-hash --flat --type sha256 case.nar)" ]
```
Comment (SuperSandro2000) [inline tests/functional/case-hack.sh:22]:
```suggestion
nix-store "${opts[@]}" --restore "$TEST_ROOT/case" < case.nar
# shellcheck disable=2086
nix-store "${opts[@]}" --dump "$TEST_ROOT/case" > "$TEST_ROOT/case.nar"
cmp case.nar "$TEST_ROOT/case.nar"
[ "$(nix-hash "${opts[@]}" --type sha256 $TEST_ROOT/case)" = "$(nix-hash --flat --type sha256 case.nar)" ]
```

File: tests/functional/case-hack.sh [hunk 15]
```diff
@@ -4,18 +4,25 @@ source common.sh
...
 
 # Check whether we detect true collisions (e.g. those remaining after
 # removal of the suffix).
 touch "$TEST_ROOT/case/xt_CONNMARK.h~nix~case~hack~3"
+# shellcheck disable=2086
 (! nix-store $opts --dump $TEST_ROOT/case > /dev/null)
```
Comment (SuperSandro2000) [inline tests/functional/case-hack.sh:28]:
```suggestion
(! nix-store "${opts[@]}" --dump $TEST_ROOT/case > /dev/null)
```

File: tests/functional/fetchTree-file.sh [hunk 16]
```diff
@@ -90,7 +90,7 @@ EOF
 EOF
 
     # Test tarball URLs on the command line.
-    [[ $(nix flake metadata --json file://$PWD/test_input_no_ext | jq -r .resolved.type) = tarball ]]
+    [[ $(nix flake metadata --json file://"$PWD"/test_input_no_ext | jq -r .resolved.type) = tarball ]]
```
Comment (SuperSandro2000) [inline tests/functional/fetchTree-file.sh:93]:
```suggestion
    [[ $(nix flake metadata --json "file://$PWD/test_input_no_ext" | jq -r .resolved.type) = tarball ]]
```
eg

File: tests/functional/flakes/absolute-attr-paths.sh (hunk 2 above)
Comment (SuperSandro2000) [reply tests/functional/flakes/absolute-attr-paths.sh:19]: Yes, the hashtag is only a comment if there is a space before.

I don't think there is an autofixer in shellcheck itself.

Also I think comment rendering could be fixed by quoting the entire string, but that's neglible. 

```suggestion
[ "$(nix eval --impure --json "$flake1Dir#.x")" -eq 1 ]
[ "$(nix eval --impure --json "$flake1Dir#x")" -eq 2 ]
[ "$(nix eval --impure --json "$flake1Dir#.packages.$system.x")" -eq 2 ]
```

**Review by SuperSandro2000 — COMMENTED** [review]
I tried adding not to many comments, especially repeating ones.

Since most of this is regarding quoting, there are not to many traps, unless we quote something which was used as an array and now is mangled into one string but we should notice that when the tests fail.

File: tests/functional/flakes/absolute-attr-paths.sh (hunk 2 above)
Comment (SkamDart) [reply tests/functional/flakes/absolute-attr-paths.sh:19]: Most of the commits were the `shellcheck -p diff | git apply`. The only exceptions were adding a shebang to the top of a file or disabling a lint to ignore a false positive

**Comment by Ericson2314** [comment]
I opened #2996 to get some upstream advice with regards to the quotation issue.

File: tests/functional/ca/build-cache.sh (hunk 12 above)
Comment (Ericson2314) [reply tests/functional/ca/build-cache.sh]: Fixed now, See the `.shellcheck` change

**Pull request merged**